Add unix socket support

clarfonthey opened the pull request:
Enables listening on unix sockets by specifying a file path for the bind address. Fixes #542.

- [x] I have signed the [CLA](https://phabricator.write.as/L1)

thebaer commented:
Great addition -- thanks for submitting this! I'll test as soon as I can. In the meantime, anyone else should feel free to review!

clarfonthey commented:
I did run the automated tests and everything passed, although I haven't gotten around to testing the actual server yet. Technically other options for unix sockets could be added, but I figured I'd choose the option with the minimum surface area for now since I'm not sure how much that's desired.

Will add some extra remarks as needed once I get around to that.

clarfonthey commented:
Okay so, I finally got around to testing this, noticed a few issues, and fixed them. Notable changes:

1. Error handling in golang is awful and I accidentally shadowed `err`, which is now fixed
2. The server needs to delete any leftover socket files in order to actually listen on the socket; if it doesn't have permission to do this, it will just throw an error
3. I found out that `http.Serve(listener, r)` exists so I did that instead of `(&http.Server{Handler: r}).Serve(listener)`

thebaer commented:
I went ahead and cleaned the code up to be a little more idiomatic, and made it so the socket file is deleted on server shutdown.

I probably undid your error-naming changes -- I'm not sure what the error shadowing issue was that you mentioned, since past code was overwritten by force-push. But let me know if this causes any issues! Otherwise I tested and it works well, so I'll merge this soon.

clarfonthey commented:
I don't remember what the error change was, but I was particularly frustrated with the compiler in that moment and the brute force approach worked. So, if your code works, that's good.

I didn't remove the socket initially since I figured it didn't matter much since it'd be overwritten on the next time it starts, but it is an improvement. Thank you.

Everything looks good to me. Appreciate you tidying this up for me.

thebaer commented:
Merging this now. Thanks again for working on this!
